How to call static methods?

Hallo,

I defined a static method on my Typegoose model and then tried to use the model within a service.
The TypeScript compiler issues an error, though:

> error TS2339: Property 'findAll' does not exist on type 'ModelType<DBShipping>'.

How can I solve this?
Here's my code:

```
export class DBShipping extends Typegoose {
	@staticMethod
	static findAll(this: ModelType<DBShipping> & typeof DBShipping, userId: UserId) {
		....
	}
        ....
}


constructor(@InjectModel(DBShipping) private shippingModel: ModelType<DBShipping>) {
	}

....
const shippings = await this.shippingModel.findAll(context.user.id);
```

Thanks!

--- a/README.md
+++ b/README.md
@@ -83,6 +83,26 @@ export class CatsService {
 }
 ```
 
+### Mongoose Schema Options
+
+To add custom [mongoose schema options](http://mongoosejs.com/docs/guide.html#options)
+you can simply change `Typegoose.forFeature` to the following format:
+
+```typescript
+@Module({
+  imports: [
+    TypegooseModule.forFeature({
+      typegooseClass: Cat,
+      schemaOptions: {
+        collection: 'ADifferentCollectionNameForCats'
+      }
+    })
+  ]
+})
+export class CatsModule {}
+```
+
+
 ## Requirements
 
  1. Typegoose +5.2.1

--- a/package.json
+++ b/package.json
@@ -25,7 +25,8 @@
     "typegoose": "~5.2.1"
   },
   "dependencies": {
-    "reflect-metadata": "^0.1.12"
+    "reflect-metadata": "^0.1.12",
+    "is-class": "latest"
   },
   "devDependencies": {
     "mongoose": "~5.1.1",

--- a/package.json
+++ b/package.json
@@ -61,6 +61,7 @@
       "^.+\\.(t|j)s$": "ts-jest"
     },
     "coverageDirectory": "../coverage",
-    "resetMocks": true
+    "resetMocks": true,
+    "testEnvironment": "node"
   }
 }

--- a/src/typegoose.module.spec.ts
+++ b/src/typegoose.module.spec.ts
@@ -48,22 +48,53 @@ describe('TypegooseModule', () => {
   });
 
   describe('forFeature', () => {
-    it('should return module that exports providers for Models', () => {
-      jest.spyOn(createProviders, 'createTypegooseProviders')
-        .mockImplementation(modelArray => modelArray.map(model => model.name));
-
-      const module = TypegooseModule.forFeature(MockTask, MockUser);
-
-      const expectedProviders = [
-        'MockTask',
-        'MockUser'
+    let models, convertedModels;
+    beforeEach(() => {
+      models = [
+        MockTask,
+        {
+          typegooseClass: MockUser,
+          schemaOptions: {
+            collection: 'differentCollectionNameUser'
+          }
+        }
       ];
 
+      let count = -1;
+      convertedModels = [
+        'convertedTask',
+        'convertedUser'
+      ];
+
+      jest.spyOn(createProviders, 'convertToTypegooseClassWithOptions')
+        .mockImplementation(() => {
+          count += 1;
+          return convertedModels[count];
+        });
+
+      jest.spyOn(createProviders, 'createTypegooseProviders')
+        .mockReturnValue('createdProviders');
+    });
+
+    it('should return module that exports providers for Models', () => {
+      const module = TypegooseModule.forFeature(...models);
+
+      const expectedProviders = 'createdProviders';
+
+      expect(createProviders.convertToTypegooseClassWithOptions).toHaveBeenCalledWith(MockTask);
+      expect(createProviders.convertToTypegooseClassWithOptions).toHaveBeenCalledWith({
+        typegooseClass: MockUser,
+        schemaOptions: {
+          collection: 'differentCollectionNameUser'
+        }
+      });
+
+      expect(createProviders.createTypegooseProviders).toHaveBeenCalledWith(convertedModels);
       expect(module).toEqual({
         module: TypegooseModule,
         providers: expectedProviders,
         exports: expectedProviders
       });
     });
-  })
+  });
 });

--- a/src/typegoose.module.ts
+++ b/src/typegoose.module.ts
@@ -1,6 +1,11 @@
 import { DynamicModule, Module } from '@nestjs/common';
 import { TypegooseCoreModule } from './typegoose-core.module';
-import { createTypegooseProviders } from './typegoose.providers';
+
+import {
+  convertToTypegooseClassWithOptions,
+  createTypegooseProviders,
+  TypegooseClassWithOptions
+} from './typegoose.providers';
 import { TypegooseClass } from './typegoose-class.interface';
 import { ConnectionOptions } from 'mongoose';
 
@@ -16,8 +21,9 @@ export class TypegooseModule {
     };
   }
 
-  static forFeature(...models: TypegooseClass<any>[]): DynamicModule {
-    const providers = createTypegooseProviders(models);
+  static forFeature(...models: (TypegooseClass<any> | TypegooseClassWithOptions)[]): DynamicModule {
+    const convertedModels = models.map(model => convertToTypegooseClassWithOptions(model));
+    const providers = createTypegooseProviders(convertedModels);
     return {
       module: TypegooseModule,
       providers,

--- a/src/typegoose.providers.spec.ts
+++ b/src/typegoose.providers.spec.ts
@@ -1,5 +1,5 @@
 import { prop, Typegoose } from 'typegoose';
-import { createTypegooseProviders } from './typegoose.providers';
+import { convertToTypegooseClassWithOptions, createTypegooseProviders } from './typegoose.providers';
 import * as mongoose from 'mongoose';
 import { Connection } from 'mongoose';
 import { Mockgoose } from 'mockgoose';
@@ -38,13 +38,59 @@ describe('createTypegooseProviders', () => {
     await mongoose.connection.close();
   });
 
+  describe('setModelForClass', () => {
+    let mockSetModel, MockTypegooseClass, mockConnection, schemaOptions, provider;
+    beforeEach(() => {
+      mockSetModel = jest.fn();
+      MockTypegooseClass = jest.fn().mockImplementation(() => ({
+        setModelForClass: mockSetModel
+      }));
+      mockConnection = jest.fn() as any;
+
+      schemaOptions = {
+        collection: 'newCollectionName'
+      };
+
+      const models = [
+        {
+          typegooseClass: MockTypegooseClass,
+          schemaOptions
+        }
+      ];
+
+      ([ provider ] = createTypegooseProviders(models));
+      provider.useFactory(mockConnection);
+    });
+
+    it('should setup the database model', () => {
+      expect(mockSetModel).toHaveBeenCalled();
+    });
+
+    it('should use existing connection from DbConnectionToken', () => {
+      expect(mockSetModel.mock.calls[0][1]).toEqual(expect.objectContaining({
+        existingConnection: mockConnection
+      }));
+    });
+
+    it('should forward schema options to typegoose', () => {
+      expect(mockSetModel.mock.calls[0][1]).toEqual(expect.objectContaining({
+        schemaOptions
+      }));
+    });
+  });
+
+
   it('should create typegoose providers from models', () => {
 
     jest.setTimeout(30000);
 
     const models = [
-      MockUser,
-      MockTask
+      {
+        typegooseClass: MockUser
+      },
+      {
+        typegooseClass: MockTask
+      }
     ];
 
     const providers = createTypegooseProviders(models);
@@ -78,4 +124,36 @@ describe('createTypegooseProviders', () => {
   afterAll(() => {
     connection.close();
   })
+});
+
+class MockTypegooseClass {}
+
+describe('convertToTypegooseClassWithOptions', () => {
+  it('returns model as typegooseClass if it is just a class', () => {
+    expect(convertToTypegooseClassWithOptions(MockTypegooseClass)).toEqual({
+      typegooseClass: MockTypegooseClass
+    });
+  });
+  it('returns model and schemaOptions if it is a TypegooseClassWithOptions', () => {
+    const options = {
+      collection: 'differentName'
+    };
+
+    const expected = {
+      typegooseClass: MockTypegooseClass,
+      schemaOptions: options
+    };
+
+    expect(convertToTypegooseClassWithOptions(expected)).toEqual(expected);
+  });
+  it('throws error is not a class or not a TypegooseClassWithOptions', () => {
+    const handler = () => {
+      expect(convertToTypegooseClassWithOptions({
+        // @ts-ignore
+        something: 'different'
+      }))
+    };
+
+    expect(handler).toThrowErrorMatchingSnapshot();
+  });
 });

--- a/src/typegoose.providers.ts
+++ b/src/typegoose.providers.ts
@@ -1,11 +1,35 @@
-import {getModelToken} from './typegoose.utils';
-import {TypegooseClass} from './typegoose-class.interface';
-import {Connection} from 'mongoose';
+import { getModelToken } from './typegoose.utils';
+import { TypegooseClass } from './typegoose-class.interface';
+import { Connection, SchemaOptions } from 'mongoose';
+import * as isClass from 'is-class';
 
-export function createTypegooseProviders(models: TypegooseClass<any>[] = []) {
-  return models.map((typegooseClass: TypegooseClass<any>) => ({
+export type TypegooseClassWithOptions = {
+  typegooseClass: TypegooseClass<any>,
+  schemaOptions?: SchemaOptions
+}
+
+export const isTypegooseClassWithOptions = (item): item is TypegooseClassWithOptions =>
+  isClass(item.typegooseClass);
+
+export const convertToTypegooseClassWithOptions = (item: TypegooseClass<any> | TypegooseClassWithOptions): TypegooseClassWithOptions => {
+  if (isClass(item)) {
+    return {
+      typegooseClass: item as TypegooseClass<any>
+    };
+  } else if (isTypegooseClassWithOptions(item)) {
+    return item;
+  }
+
+  throw new Error('Invalid model object');
+};
+
+export function createTypegooseProviders(models: TypegooseClassWithOptions[] = []) {
+  return models.map(({ typegooseClass, schemaOptions = {} }) => ({
     provide: getModelToken(typegooseClass.name),
-    useFactory: (connection: Connection) => new typegooseClass().getModelForClass(typegooseClass, {existingConnection: connection}),
+    useFactory: (connection: Connection) => new typegooseClass().setModelForClass(typegooseClass, {
+      existingConnection: connection,
+      schemaOptions
+    }),
     inject: ['DbConnectionToken']
   }));
 }

